azurelinuxagent/common/protocol/wire.py

```diff
@@ -774,42 +710,25 @@ def _update_from_goal_state(self, refresh_type):
         """
         Fetches a new goal state and updates the internal state of the WireClient according to the requested 'refresh_type'
         """
-        uri = GOAL_STATE_URI.format(self.get_endpoint())
-
         max_retry = 3
 
         for retry in range(1, max_retry + 1):
             try:
-                new_goal_state_xml = self.fetch_config(uri, self.get_header())
-                new_goal_state = GoalState(new_goal_state_xml)
-
-                def update_host_plugin():
-                    if self._host_plugin is not None:
-                        self._host_plugin.update_container_id(new_goal_state.container_id)
-                        self._host_plugin.update_role_config_name(new_goal_state.role_config_name)
-
                 if refresh_type == WireClient._UpdateType.HostPlugin:
-                    update_host_plugin()
+                    goal_state = GoalState.fetch_goal_state(self)
+                    self._update_host_plugin(goal_state.container_id, goal_state.role_config_name)
                     return
 
-                if self._goal_state is None or refresh_type == WireClient._UpdateType.GoalStateForced or \
-                        new_goal_state.incarnation != self._goal_state.incarnation:
-                    def save_goal_state(incarnation, xml_text):
-                        file_name = GOAL_STATE_FILE_NAME.format(incarnation)
-                        goal_state_file = os.path.join(conf.get_lib_dir(), file_name)
-                        self.save_cache(goal_state_file, xml_text)
-
-                    self.goal_state_flusher.flush(datetime.utcnow())
-                    self._set_goal_state(new_goal_state)
-                    save_goal_state(new_goal_state.incarnation, new_goal_state_xml)
-                    self.update_hosting_env(new_goal_state)
-                    self.update_shared_conf(new_goal_state)
-                    self.update_certs(new_goal_state)
-                    self.update_ext_conf(new_goal_state)
-                    self.update_remote_access_conf(new_goal_state)
-                    self.save_cache(os.path.join(conf.get_lib_dir(), INCARNATION_FILE_NAME), new_goal_state.incarnation)
-
-                    update_host_plugin()
+                if self._goal_state is None or refresh_type == WireClient._UpdateType.GoalStateForced:
+                    new_goal_state = GoalState.fetch_full_goal_state(self)
+                else:
+                    new_goal_state = GoalState.fetch_full_goal_state_if_incarnation_different_than(self, self._goal_state.incarnation)
+
+                if new_goal_state is not None:
+                    self._goal_state = new_goal_state
```

Comment:
this fixes the original issue: the entire goal state is updated or no updates are done